src/kernel/binary_reduce.cc

```diff
@@ -0,0 +1,506 @@
+    const int dr = (rhs->ndim - 1 - j < 1)? 1 : rhs->shape[rhs->ndim - 1 - j];
+    if (dl != dr) {
+      if (dl != 1 && dr != 1) {
+        LOG(FATAL) << "Invalid broadcasting between feature shapes "
+          << ShapeString(lhs) << " and " << ShapeString(rhs);
+      }
+      if (accum != 0) {
+        ret.lhs_shape.push_back(accum);
+        ret.rhs_shape.push_back(accum);
+        ret.out_shape.push_back(accum);
+        accum = 0;
+      }
+      ret.lhs_shape.push_back(dl);
+      ret.rhs_shape.push_back(dr);
+      ret.out_shape.push_back(std::max(dl, dr));
+    } else {
+      if (accum == 0) {
+        accum = dl;
+      } else {
+        accum *= dl;
+      }
+    }
+    ret.real_out_shape.push_back(std::max(dl, dr));
+  }
+  if (accum != 0) {
+    ret.lhs_shape.push_back(accum);
+    ret.rhs_shape.push_back(accum);
+    ret.out_shape.push_back(accum);
+    accum = 0;
+  }
+  std::reverse(ret.real_out_shape.begin(), ret.real_out_shape.end());
+  std::reverse(ret.lhs_shape.begin(), ret.lhs_shape.end());
+  std::reverse(ret.rhs_shape.begin(), ret.rhs_shape.end());
+  std::reverse(ret.out_shape.begin(), ret.out_shape.end());
+  // stride
+  ret.lhs_stride = ComputeStride(ret.lhs_shape);
+  ret.rhs_stride = ComputeStride(ret.rhs_shape);
+  ret.out_stride = ComputeStride(ret.out_shape);
+  return ret;
+}
+
+std::string IdArrayToStr(IdArray arr) {
+  arr = arr.CopyTo(DLContext{kDLCPU, 0});
+  int64_t len = arr->shape[0];
+  std::ostringstream oss;
+  oss << "(" << len << ")[";
+  if (arr->dtype.bits == 32) {
+    int32_t* data = static_cast<int32_t*>(arr->data);
+    for (int64_t i = 0; i < len; ++i) {
+      oss << data[i] << " ";
+    }
+  } else {
+    int64_t* data = static_cast<int64_t*>(arr->data);
+    for (int64_t i = 0; i < len; ++i) {
+      oss << data[i] << " ";
+    }
+  }
+  oss << "]";
+  return oss.str();
+}
+
+inline void CheckCtx(
+    const DLContext& ctx,
+    const std::vector<NDArray>& arrays,
+    const std::vector<std::string>& names) {
+  for (size_t i = 0; i < arrays.size(); ++i) {
+    if (utils::IsNoneArray(arrays[i]))
+      continue;
+    CHECK_EQ(ctx, arrays[i]->ctx)
+      << "Expected device context " << ctx << ". But got "
+      << arrays[i]->ctx << " for " << names[i] << ".";
+  }
+}
+
+inline void CheckIdArray(
+    const uint8_t bits,
+    const std::vector<NDArray>& arrays,
+    const std::vector<std::string>& names) {
+  for (size_t i = 0; i < arrays.size(); ++i) {
+    if (utils::IsNoneArray(arrays[i]))
+      continue;
+    CHECK(arrays[i]->dtype.code == kDLInt);
+    CHECK_EQ(arrays[i]->ndim, 1);
+    CHECK_EQ(bits, arrays[i]->dtype.bits)
+      << "Expected " << bits << " integer array. But got "
+      << arrays[i]->dtype.bits << " for " << names[i] << ".";
+  }
+}
+
+// Return true if the operator is commutative and lhs and rhs need
+// to be switched. For example, Add(kDst, kSrc) needs to be changed
+// to Add(kSrc, kDst)
```

Comment:
i don't understand. why we need to switch "Add(kDst, kSrc)" to "Add(kSrc, kDst)".

Reply:
This is to speedup the compilation. We don't want to generate kernels for both Add(kDst, kSrc) and Add(kSrc, kDst). They are the same.

Reply:
More comments. Pls checkout.